### ISSUE TITLE
FmtToIoWriter flushes on every tiny string written

### ISSUE THREAD

**Issue by emk** [dtolnay/serde-yaml#28]
Thank you for the serde_yaml support! We're making heavy use of this in [cage](https://github.com/faradayio/cage).

[There's a call to `flush` in `FmtToIoWriter`](https://github.com/dtolnay/serde-yaml/blob/b8bf364ae4ed6323a8154feb859ea9327d473a8b/yaml/src/ser.rs#L417) that slows down YAML output by making a large number of system calls:

Sample strace output:

```
write(3, "\"", 1)                       = 1
write(3, "\n", 1)                       = 1
write(3, " ", 1)                        = 1
write(3, " ", 1)                        = 1
write(3, " ", 1)                        = 1
write(3, " ", 1)                        = 1
write(3, " ", 1)                        = 1
write(3, " ", 1)                        = 1
write(3, "\"", 1)                       = 1
write(3, "ENV_NAME", 8)                 = 8
write(3, "\"", 1)                       = 1
write(3, ": ", 2)                       = 2
write(3, "\"", 1)                       = 1
write(3, "development", 11)             = 11
write(3, "\"", 1)                       = 1
write(3, "\n", 1)                       = 1
write(3, " ", 1)                        = 1
write(3, " ", 1)                        = 1
write(3, " ", 1)                        = 1
write(3, " ", 1)                        = 1
write(3, " ", 1)                        = 1
write(3, " ", 1)                        = 1
write(3, "\"", 1)                       = 1
```

Since we do a large amount of YAML output, these syscalls are taking up a large amount of our program's runtime.

I'd be happy to try and prepare PR if you're interested.

**Comment by dtolnay** [comment]
Good call, I agree the flush shouldn't be there. I fixed this in a nonbreaking way in https://github.com/dtolnay/serde-yaml/commit/5df9007714aba3c717d30e859e8feb95bc13a89c and released [0.4.1](https://crates.io/crates/serde_yaml/0.4.1). I filed #29 to remove the flush altogether because serde_json has been fine without one. Is it still going to be a bottleneck for you with a single flush? If so, I can get the ball rolling on 0.5.0.

**Comment by emk** [comment]
Excellent! Thank you for the fast response.

I'm seeing a factor of **20** speedup for `cage` runs. This is a huge improvement. And I don't think that final `flush` is hurting things one way or the other, so we're good there.

The massive lists of tiny writes now happen as large single blocks, and the `flush` happens purely at the Rust level, not the OS level:

```
write(3, "---\n\"services\": \n  \"strider\": \n "..., 4175) = 4175
close(3)                                = 0
```

Thank you very much!
